sdk-lib/src/main/rust/lib.rs

```diff
@@ -226,40 +219,30 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_tool_DerivationTool_deriveU
             return Err(format_err!("accounts argument must be greater than zero"));
         };
 
-        let extfvks: Vec<_> = (0..accounts)
-            .map(|account| {
-                encode_extended_full_viewing_key(
-                    network.hrp_sapling_extended_full_viewing_key(),
-                    &ExtendedFullViewingKey::from(&spending_key(
-                        &seed,
-                        network.coin_type(),
-                        AccountId(account),
-                    )),
-                )
-            })
-            .collect();
-
-        let extpubs: Vec<_> = (0..accounts)
+        let ufvks: Vec<_> = (0..accounts)
             .map(|account| {
-                let pk =
-                    derive_public_key_from_seed(&network, &seed, AccountId(account), 0).unwrap();
-                hex::encode(&pk.serialize())
+                let account_id = AccountId::from(account);
+                UnifiedSpendingKey::from_seed(&network, &seed, account_id)
+                    .map_err(|e| {
+                        format_err!("error generating unified spending key from seed: {:?}", e)
+                    })
+                    .map(|usk| usk.to_unified_full_viewing_key().encode(&network))
             })
-            .collect();
+            .collect::<Result<_, _>>()?;
 
-        Ok(utils::rust_vec_to_java_2d(
+        Ok(utils::rust_vec_to_java(
             &env,
-            extfvks,
-            extpubs,
-            |env, extfvkstr| env.new_string(extfvkstr),
+            ufvks,
+            "java/lang/String",
```

Comment:
Abstractions leaking, leaking everywhere. (Don't mind me. Just grumbling at the state of software infrastructure.)